File: src/java/gov/usgs/geomag/io/TimeseriesFactory.java
package gov.usgs.geomag.io;

import java.io.IOException;
import java.util.Date;

public interface TimeseriesFactory {

	/**
	 * Retrieve timeseries data for a specified observatory and component.
	 *
	 * @param observatory the observatory to fetch.
	 * @param component the component to fetch.
	 * @param starttime the time of the first sample.
	 * @param endtime the time of the last sample.
	 */
	public Timeseries getTimeseries(String observatory, String component,
			Date starttime, Date endtime) throws IOException;

}

File: src/python/geomag/io/TimeseriesFactory.py
"""Abstract Timeseries Factory Interface."""


class TimeseriesFactory(object):
    """Base class for timeseries factories."""

    def get_timeseries(self, observatory, starttime, endtime,
            channels=('H', 'D', 'Z', 'F'),
            type='variation', interval='minute'):
        """Get timeseries data.

        Support for specific channels, types, and intervals varies
        between factory and observatory.  Subclasses should raise
        TimeseriesFactoryException if the data is not available, or
        if an error occurs accessing data.

        Parameters
        ----------
        observatory : str
            observatory code, usually 3 characters.
        starttime : UTCDateTime
            time of first sample in timeseries.
        endtime : UTCDateTime
            time of last sample in timeseries.
        channels : array_like
            list of channels to load.
        type : {'definitive', 'provisional', 'quasi-definitive', 'variation'}
            data type.
        interval : {'daily', 'hourly', 'minute', 'monthly', 'second'}
            data interval.

        Returns
        -------
        obspy.core.Stream
            stream containing traces for requested timeseries.

        Raises
        ------
        TimeseriesFactoryException
            if any parameters are unsupported, or errors occur loading data.
        """
        raise NotImplementedError('"get_timeseries" not implemented')

    def put_timeseries(self, timeseries, channels):
        """Store timeseries data.

        Parameters
        ----------
        timeseries : obspy.core.Stream
            stream containing traces to store.
        channels : list
            list of channels to store.

        Raises
        ------
        TimeseriesFactoryException
            if any errors occur.
        """
        raise NotImplementedError('"put_timeseries" not implemented')
